Remove deprecated JetCorrector and related files

wddgit opened the pull request:
#### PR description:

Remove the deprecated ::JetCorrector class from CMSSW and other related files with it. The migration from ::JetCorrector to reco::JetCorrector started in 2014 and it is time to finally complete it. For the most part this PR removes unused dead code and simplifies things. It makes it impossible for someone to use the deprecated code and obvious failures will occur if they try. It reduces the amount of code we need to maintain and support.

Almost this entire PR consists of deletions. There is one exception to that. The exception is in RecoHI/HiJetAlgos/python/hiCaloJetsForTrk_cff.py. I migrated that file from using the old JetCorrector type to using the new reco::JetCorrector. Somehow that one was missed earlier.

There has been a long effort to migrate from the old JetCorrectors to the new JetCorrectors that has been in progress since 2014. Volker Adler did most of this migration long ago. I've recently migrated all the remaining C++ code. I've also migrated all the Python configuration code that is used in existing tests that I can find. And I will help if merging this PR reveals more cases of code in CMSSW that needs migration. There are 4 PRs that are submitted but not merged yet. These need to be merged before this PR or there will be compilation errors. These are #39875 #39864 #39847 and #39804.

I also deleted the obsolete central cff configuration files commonly imported to use the old JetCorrector class.

There are dozens of other configuration files that still reference the old cff files deleted by this PR. Those will be broken by this PR, but they are all old and look to be old obsolete code that is already broken. It is not worth the effort to deal with them at this point. If someone tries to run them they will fail with an obvious exception and then they can migrated to use the new JetCorrector class. There is a good chance none of them will ever be used again.

#### PR validation:

Relies on existing tests. Errors caused by this should result in obvious exceptions.

The one exception is the change in hiCaloJetsForTrk_cff.py. That is tested by runTheMatrix.py test 140.53. That is how I noticed it still needed to be migrated.

If other failures show up, I will be available to help migrate the affected code.

wddgit committed:
Remove deprecated JetCorrector and related files

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39953/32883

- This PR adds an extra 20KB to repository

wddgit commented:
please test with #39875,#39864,#39847,#39804 

FYI @Dr15Jones @makortel

makortel commented on JetMETCorrections/Objects/interface/JetCorrector.h:
```diff
@@ -1,64 +1 @@
-inline bool JetCorrector::vectorialCorrection() const { return false; }
-
-#endif
+#error, obsolete header, use reco::JetCorrector in JetMETCorrections/JetCorrector/interface/JetCorrector.h instead
```
Should this be 
```suggestion
#error "obsolete header, use reco::JetCorrector in JetMETCorrections/JetCorrector/interface/JetCorrector.h instead"
```
?

wddgit replied:
I will fix that as soon as the tests finish. Thanks. No idea why I put a comma there...

cmsbuild commented:
-1

**Failed Tests**:   HeaderConsistency 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6983d/28715/summary.html
**COMMIT**: 2ad4017ea8e3c65c9209fc21c0b7b3af01e2fccb
**CMSSW**: CMSSW_12_6_X_2022-11-01-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39953/28715/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 534.0 step 1
536.0 step 1
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416398
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416364
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-01-1100+c6983d/53850/triggerResults): no differences found

makortel commented:
> -1
> 
> **Failed Tests**: HeaderConsistency

Note that the header consistency check failure is caused by the `#error`, i.e. it is expected.

wddgit commented:
please test with #39875,#39864,#39847,#39804

with fix to "#error" line

cmsbuild commented:
Pull request #39953 was updated. @swertz, @vlimant, @micsucmed, @emanueleusai, @ahmad3213, @jfernan2, @clacaputo, @syuvivida, @pmandrik, @mandrenguyen, @rvenditti can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**:   HeaderConsistency RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6983d/28743/summary.html
**COMMIT**: f3fd3e173fffff3b2c97b2cae30d2eefa3ba7581
**CMSSW**: CMSSW_12_6_X_2022-11-02-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39953/28743/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416398
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416367
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-02-1100+c6983d/53873/triggerResults): no differences found

wddgit commented:
@smuzaffar Is there a way to make "scram b check-headers" ignore the file where I added "#error"? Would it be better if I just deleted the file instead? The only value it adds is a small message, but usage of the header is already completely removed from CMSSW, so it is not that important. I would hate to have this error appearing every time we have a PR including this package.

wddgit commented:
In my local working area, I ran the the full complete set of runTheMatrix.py tests, about 1000 of them. All passed except there were 2 unrelated failures ("TritonService: Starting the fallback server failed with exit code 65280" which has nothing to do with JetCorrectors).

I'm leaning towards just deleting the header file instead of putting "#error" in there... And still waiting on the 4 PRs this depends on. Please don't merge this PR until those are also merged.

wddgit committed:
Remove JetCorrector.h

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-39953/32916

- This PR adds an extra 8KB to repository

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6983d/28789/summary.html
**COMMIT**: 44852c6f38e125b178a7204c0ffc3e6051cc956d
**CMSSW**: CMSSW_12_6_X_2022-11-04-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39953/28789/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestDQMOfflineConfiguration_150 had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416402
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416377
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-04-1100+c6983d/53903/triggerResults): no differences found

wddgit commented:
please test

I have not seen this unit test failure before, TestDQMOfflineConfiguration_150. It not obvious from the log what the problem is. Maybe a glitch. Try again.

wddgit commented:
abort test

wddgit commented:
please test with #39875, #39864, #39847, #39804

I have not seen this unit test failure before, TestDQMOfflineConfiguration_150. It not obvious from the log what the problem is. Maybe a glitch. Try again.

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6983d/28797/summary.html
**COMMIT**: 44852c6f38e125b178a7204c0ffc3e6051cc956d
**CMSSW**: CMSSW_12_6_X_2022-11-04-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39953/28797/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416402
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416374
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-04-1100+c6983d/53909/triggerResults): no differences found

wddgit commented:
please test with #39875,#39864,#39847,#39804

Try again, unit tests passed last time, but now relVal tests which passed before timed out, does not look related to my changes although it is hard to understand why they failed

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6983d/28803/summary.html
**COMMIT**: 44852c6f38e125b178a7204c0ffc3e6051cc956d
**CMSSW**: CMSSW_12_6_X_2022-11-05-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/39953/28803/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 9 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3416402
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3416371
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 48 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_6_X_2022-11-05-1100+c6983d/53916/triggerResults): no differences found

emanueleusai commented:
+1

- I agree with the strategy to let the obsolete code as it is. In DQM we just contacted all PAGs and POGs to check which set of monitoring elements are actually being used. We should be able to clean up a good number of the obsolete modules in DQM/Physics once we get an answer from all the groups

swertz commented:
+xpog

clacaputo commented:
+reconstruction

- PR removes deprecated JetCorrector 
- #39875,#39864,#39847,#39804 are related to this one. These need to be merged before or there will be compilation errors
- no reco changes observed

perrotta commented:
Note for the release managers: as of now this still need https://github.com/cms-sw/cmssw/pull/39847 and https://github.com/cms-sw/cmssw/pull/39804 to be merged. This PR will follow

wddgit commented:
I closed #39804 and replaced it with #40139. So now don't merge this until both #40139 and #39847 are merged. Should I just put this on hold to reduce confusion?

perrotta commented:
> Should I just put this on hold to reduce confusion?

Yes, why not David?
I'd let you put on hold, instead of doing it myself, so that you can unhold yourself as soon as all other pieces are ready

wddgit commented:
hold

makortel commented:
hold

(seems that holding needs more power)

cmsbuild commented:
Pull request has been put on hold by  @makortel
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

wddgit commented:
please test

@makortel I think we can remove the hold. I'm rerunning the tests to see if they will pass with no other PR's included. I think it should now that all the other PRs were merged.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c6983d/29464/summary.html
**COMMIT**: 44852c6f38e125b178a7204c0ffc3e6051cc956d
**CMSSW**: CMSSW_13_0_X_2022-12-05-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/39953/29464/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 25 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3421337
 - DQMHistoTests: Total failures: 1170
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3420145
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 206 log files, 158 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-05-1100+c6983d/54523/triggerResults): no differences found

makortel commented:
The comparison differences are in 11634.7, so this looks like another symptom of https://github.com/cms-sw/cmssw/issues/39803#issuecomment-1337009470

rappoccio commented:
+1

Comparison differences are spurious.

perrotta commented:
+1

perrotta commented:
@wddgit after the merge of this PR there are unit test errors in Configuration/DataProcessing, apparently related to the relics JetCorrectionServicesAllAlgos_cff.py which is still imported in
- HeavyIonsAnalysis/Configuration/python/HI_DiJetSkim_cff.py
- HeavyIonsAnalysis/Configuration/python/analysisFilters_cff.py

In reality I see many more places in CMSSW where that config is imported, see e.g. https://cmssdt.cern.ch/lxr/search?%21v=CMSSW_13_0_X_2022-12-12-1100&_filestring=&_string=rrectionServicesAllAlgos_cff

Could you please have a look?

wddgit commented:
I'll take a look this morning.

wddgit commented:
The PR fixing the unit test is #40301, recently submitted.

I also submitted a new Issue with the list of files importing the deleted cff files. Issue #40304. Note that none of these are used in tests. As far as I know they are not used at all.

wddgit commented:
Starting with the IB after this was merged, CMSSW_13_0_X_2022-12-12-1100, the Static Analyzer test in the IB named "SA EventSetupRecord::get called" is now reporting this:

"no-package: {}"

I presume this means it found no problems. Clearing that was the main goal of this PR.

Also after the #40301 was merged last night, the unit tests all passed.

perrotta commented:
Good. Thank you @wddgit , also for the github issue
